docs/src/expressions.md

```diff
@@ -4,30 +4,202 @@ DocTestSetup = quote
+first argument is the constant term, and the remaining arguments are
+variable-coefficient pairs.
+
+```jldoctest
+model = Model()
+@variable(model, x)
+@variable(model, y)
+ex = AffExpr(-1.0, x => 2.0, y => 1.0)
+
+# output
+
+2 x + y - 1
+```
+
+### `JuMP.add_to_expression!`
+
+The fourth way to create an affine expression is by using
+[`JuMP.add_to_expression!`](@ref). Compared to the operator overloading method,
+this approach is fast, and is what the [`@expression`](@ref) macro implements
+behind-the-scenes.
 ```jldoctest
-m = Model()
-@variable(m, x)
-@variable(m, y)
-AffExpr(-1.0, x => 2.0, y => 1.0)
+model = Model()
+@variable(model, x)
+@variable(model, y)
+ex = JuMP.AffExpr(-1.0)
+JuMP.add_to_expression!(ex, 2.0, x)
+JuMP.add_to_expression!(ex, 1.0, y)
 
 # output
 
 2 x + y - 1
 ```
 
-Objective functions
--------------------
+## Quadratic expressions
 
-TODO: Describe how JuMP expressions relate to MOI functions. How to set, query,
-and modify an objective function.
+Like affine expressions, there are four ways of constructing a quadratic
+expression in JuMP: using macros, operator overloading, constructors, and via
+[`JuMP.add_to_expression!`](@ref).
+
+### Macros
+
+The [`@expression`](@ref) macro can be used to create quadratic expressions by
+including quadratic terms.
+
+```jldoctest
+model = Model()
+@variable(model, x)
+@variable(model, y)
+ex = @expression(model, x^2 + 2 * x * y + y^2 + x + y - 1)
+
+# output
+
+x² + 2 x*y + y² + x + y - 1
+```
+
+### Operator overloading
+
+Operator overloading can also be used to create quadratic expressions. The same
+performance warning (discussed in the affine expression section) applies.
+
+```jldoctest
+model = Model()
+@variable(model, x)
+@variable(model, y)
+ex = x^2 + 2 * x * y + y^2 + x + y - 1
+
+# output
+
+x² + 2 x*y + y² + x + y - 1
+```
+
+### Constructors
+
+Quadratic expressions can also be created using the `QuadExpr` constructor. The
+first argument is an affine expression, and the remaining arguments are pairs,
+where the first term is a `JuMP.UnorderedPair` and the second term is the
+coefficient.
+
+```jldoctest
+model = Model()
+@variable(model, x)
+@variable(model, y)
+aff_expr = AffExpr(-1.0, x => 1.0, y => 1.0)
+quad_expr = JuMP.QuadExpr(aff_expr,
+    JuMP.UnorderedPair(x, x) => 1.0,
+    JuMP.UnorderedPair(x, y) => 2.0,
+    JuMP.UnorderedPair(y, y) => 1.0
+)
+
+# output
+
+x² + 2 x*y + y² + x + y - 1
+```
+
+### `JuMP.add_to_expression!`
+
+Finally, [`JuMP.add_to_expression!`](@ref) can also be used to add quadratic
+terms.
+
+```jldoctest
+model = Model()
+@variable(model, x)
+@variable(model, y)
+ex = JuMP.QuadExpr(x + y - 1.0)
+JuMP.add_to_expression!(ex, 1.0, x, x)
+JuMP.add_to_expression!(ex, 2.0, x, y)
+JuMP.add_to_expression!(ex, 1.0, y, y)
+
+# output
+
+x² + 2 x*y + y² + x + y - 1
+```
+
+## Nonlinear expressions
+
+Nonlinear expressions can only be constructed using the [`@NLexpression`](@ref)
```

Comment:
And they can be used only in `@NLexpression`, `@NLobjective` and `@NLconstraint`. OTOH affine and quadratic expressions can't be used in `@NLexpression`, `@NLobjective` or `@NLconstraint`.